File: spring-email-java/src/main/java/com/github/migangqui/spring/email/service/EmailServiceImpl.java
package com.github.migangqui.spring.email.service;

import com.github.migangqui.spring.email.bean.SendEmailResult;
import com.github.migangqui.spring.email.bean.Email;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.InputStreamResource;
import org.springframework.mail.MailException;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.scheduling.annotation.Async;
import org.springframework.scheduling.annotation.AsyncResult;
import org.springframework.stereotype.Service;

import javax.mail.MessagingException;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.concurrent.Future;

@Slf4j
@RequiredArgsConstructor
public class EmailServiceImpl implements EmailService {

    private final JavaMailSender javaMailSender;

    @Override
    public SendEmailResult send(Email email) {
        SendEmailResult result;
        log.debug("Sending email");
        try {
            MimeMessage generatedMailMessage = generateMailMessage(email);
            javaMailSender.send(generatedMailMessage);
            log.debug("Email sent successfully");
            result = SendEmailResult.builder().status(200).build();
        } catch (MessagingException | MailException e) {
            log.warn("An error has ocurred sending email", e);
            result = SendEmailResult.builder()
                    .status(500).cause(e.getMessage()).exception(e)
                    .build();
        }
        return result;
    }

    @Async
    @Override
    public Future<SendEmailResult> sendAsync(Email email) {
        return new AsyncResult<>(send(email));
    }

    /* Private methods */

    private MimeMessage generateMailMessage(Email email) throws MessagingException {
        MimeMessageHelper helper = new MimeMessageHelper(javaMailSender.createMimeMessage(), email.getFile() != null);;

        helper.setFrom(new InternetAddress(email.getFrom()));
        helper.setTo(email.getTo());
        helper.setSubject(email.getSubject());
        helper.setText(email.getBody(), true);

        if (email.getFile() != null) {
            helper.addAttachment(email.getFilename(), new InputStreamResource(email.getFile()));
        }
        return helper.getMimeMessage();
    }

}

File: spring-email-kotlin/src/main/kotlin/com/github/migangqui/spring/email/service/EmailServiceImpl.kt
package com.github.migangqui.spring.email.service

import com.github.migangqui.spring.email.bean.Email
import com.github.migangqui.spring.email.bean.SendEmailResult
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.io.InputStreamResource
import org.springframework.mail.MailException
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessageHelper
import org.springframework.scheduling.annotation.Async
import org.springframework.scheduling.annotation.AsyncResult
import org.springframework.stereotype.Service
import java.util.concurrent.Future
import javax.mail.MessagingException
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage

@Service
class EmailServiceImpl(private val javaMailSender: JavaMailSender) : EmailService {

    private val log = KotlinLogging.logger {}

    override fun send(email: Email): SendEmailResult {
        log.info("Sending email to {}", email.to)
        return try {
            val generatedMailMessage = generateMailMessage(email)
            javaMailSender.send(generatedMailMessage)
            log.debug("Email sent successfully to {}", email.to)
            SendEmailResult(status = 200)
        } catch (e: MessagingException) {
            log.warn("An error has ocurred sending email", e)
            SendEmailResult(status = 500, cause = e.message, exception = e)
        } catch (e: MailException) {
            log.warn("An error has ocurred sending email", e)
            SendEmailResult(status = 500, cause = e.message, exception = e)
        }
    }

    @Async
    override fun sendAsync(email: Email): Future<SendEmailResult> {
        return AsyncResult(send(email))
    }

    /* Private methods */

    @Throws(MessagingException::class)
    private fun generateMailMessage(email: Email): MimeMessage {
        val generateMailMessage = javaMailSender.createMimeMessage()
        val helper = MimeMessageHelper(generateMailMessage, email.file != null)
        helper.setFrom(InternetAddress(email.from!!))
        helper.setTo(email.to!!)
        helper.setSubject(email.subject!!)
        helper.setText(email.body!!, true)
        if (email.file != null && email.filename != null) {
            helper.addAttachment(email.filename, InputStreamResource(email.file))
        }
        return helper.mimeMessage
    }

}